File: ts/packages/luzid-grpc-client/src/client/store.ts
import { createClient } from '@luzid/grpc-connection'
import type { Channel } from '@luzid/grpc-connection'

import {
  StoreGetAccountDataRequest,
  StoreGetAccountDataResponse,
  StoreGetAccountDataServiceDefinition,
  StoreGetAccountDataServiceClient,
  StoreGetDiffedAccountUpdateRequest,
  StoreGetDiffedAccountUpdateResponse,
  StoreGetDiffedAccountUpdateServiceDefinition,
  StoreGetDiffedAccountUpdateServiceClient,
} from '@luzid/grpc'

export {
  RpcStoreAccountData,
  StoreGetAccountDataRequest,
  StoreGetAccountDataResponse,
  StoreGetDiffedAccountUpdateRequest,
  StoreGetDiffedAccountUpdateResponse,
} from '@luzid/grpc'

// -----------------
// StoreGetAccountDataClient
// -----------------
class StoreGetAccountDataClient {
  private readonly client: StoreGetAccountDataServiceClient

  constructor(channel: Channel) {
    this.client = createClient(StoreGetAccountDataServiceDefinition, channel)
  }

  getAccountData(
    request: StoreGetAccountDataRequest
  ): Promise<StoreGetAccountDataResponse> {
    return this.client.getAccountData(request)
  }
}

// -----------------
// StoreGetDiffedAccountUpdateClient
// -----------------
class StoreGetDiffedAccountUpdateClient {
  private readonly client: StoreGetDiffedAccountUpdateServiceClient

  constructor(channel: Channel) {
    this.client = createClient(
      StoreGetDiffedAccountUpdateServiceDefinition,
      channel
    )
  }

  getDiffedAccountUpdate(
    request: StoreGetDiffedAccountUpdateRequest
  ): Promise<StoreGetDiffedAccountUpdateResponse> {
    return this.client.getDiffedAccountUpdate(request)
  }
}

// -----------------
// Consolidated StoreClient
// -----------------
export class StoreClient {
  private readonly getAccountDataClient: StoreGetAccountDataClient
  private readonly getDiffedAccountUpdateClient: StoreGetDiffedAccountUpdateClient

  constructor(channel: Channel) {
    this.getAccountDataClient = new StoreGetAccountDataClient(channel)
    this.getDiffedAccountUpdateClient = new StoreGetDiffedAccountUpdateClient(
      channel
    )
  }

  getAccountData(
    request: StoreGetAccountDataRequest
  ): Promise<StoreGetAccountDataResponse> {
    return this.getAccountDataClient.getAccountData(request)
  }

  getDiffedAccountUpdate(
    request: StoreGetDiffedAccountUpdateRequest
  ): Promise<StoreGetDiffedAccountUpdateResponse> {
    return this.getDiffedAccountUpdateClient.getDiffedAccountUpdate(request)
  }
}

File: dart/luzid-grpc-client/lib/src/client/store.dart
import 'package:grpc/grpc_connection_interface.dart';
import 'package:luzid_grpc/luzid_grpc.dart';

// -----------------
// StoreGetAccountDataClient
// -----------------
class StoreGetAccountDataClient {
  final StoreGetAccountDataServiceClient _client;

  StoreGetAccountDataClient(ClientChannelBase channel)
      : _client = StoreGetAccountDataServiceClient(channel);

  Future<StoreGetAccountDataResponse> getAccountData(
    StoreGetAccountDataRequest request,
  ) {
    return _client.getAccountData(request);
  }
}

// -----------------
// StoreGetDiffedAccountUpdateClient
// -----------------
class StoreGetDiffedAccountUpdateClient {
  final StoreGetDiffedAccountUpdateServiceClient _client;

  StoreGetDiffedAccountUpdateClient(ClientChannelBase channel)
      : _client = StoreGetDiffedAccountUpdateServiceClient(channel);

  Future<StoreGetDiffedAccountUpdateResponse> getDiffedAccountUpdate(
    StoreGetDiffedAccountUpdateRequest request,
  ) {
    return _client.getDiffedAccountUpdate(request);
  }
}

// -----------------
// Consolidated StoreClient
// -----------------
class StoreClient {
  final StoreGetAccountDataClient _getAccountDataClient;
  final StoreGetDiffedAccountUpdateClient _getDiffedAccountUpdateClient;

  StoreClient(ClientChannelBase channel)
      : _getAccountDataClient = StoreGetAccountDataClient(channel),
        _getDiffedAccountUpdateClient =
            StoreGetDiffedAccountUpdateClient(channel);

  Future<StoreGetAccountDataResponse> getAccountData(
    StoreGetAccountDataRequest request,
  ) {
    return _getAccountDataClient.getAccountData(request);
  }

  Future<StoreGetDiffedAccountUpdateResponse> getDiffedAccountUpdate(
    StoreGetDiffedAccountUpdateRequest request,
  ) {
    return _getDiffedAccountUpdateClient.getDiffedAccountUpdate(request);
  }
}
